### DISCUSSION TITLE
validate ACK on Mirth

### DISCUSSION THREAD

**Discussion by trilochanmukkala** [nextgenhealthcare/connect#6329] [Q&A]
I have Mirth on port 8081, have my source as database reader and destination as a TCP Sender. I'm sending the messages to HL7 Soup and I can see the messages on receiving end and an ACK is generated and sent back to Mirth but on a different ephemeral port. I need to flag the message as marked based on ACK received by Mirth but unsure how to validate the ACK. I have a transformer and response script set up on the TCP Sender. attached is my response script and I keep getting 'null' response. Am I using the right approach?
[response.txt](https://github.com/user-attachments/files/17499335/response.txt)

**Comment by jonbartels** [comment]
In your response transformer the `msg` variable is the ACK.

I think you can do something like this:
```
var ackType = responseMap.get("response");

var ackType = msg['MSA']['MSA.1'].toString();

if (ackType == "AE" || ackType == "AR") {
  logger.error("Nack received:" + ackType);  
}

```

**Reply by trilochanmukkala** [reply]
I'm actually getting a response timeout. I have tried increasing the response timeout to large number and 0 (was suggested to use for infinite timeout). wireshark logs suggest that HL7 soup is sending the ACK to Mirth (version 4.5.2)
this is my current TCP sender config
![image](https://github.com/user-attachments/assets/dd5b875e-4c02-48dd-aba5-8db215bcc9f5)

**Reply by pacmano1** [reply]
An ACK would not be sent on a different port.   It's a single transaction over the same TCP dialog.  

Fix the receiving end.

**Reply by trilochanmukkala** [reply]
the HL7 soup is sending ACK to the same port

**Reply by pacmano1** [reply]
what does "and an ACK is generated and sent back to Mirth but on a different ephemeral port." mean???

**Reply by trilochanmukkala** [reply]
> what does "and an ACK is generated and sent back to Mirth but on a different ephemeral port." mean???

during my initial investigation I thought the message has been sent to a different ephemeral port but later looking at the network logs suggested that it HL7 soup is sending ACK to the same port it received the message from

**Reply by pacmano1** [reply]
Something seems off with your HL7 Soup set up.

Also on the response:

1. If you send a message from Mirth and the other side NACKs it, mirth will ERROR the message.  
2. If you send a message and the other side ACKs it properly, Mirth will mark the message as SENT.

The problem with bullet 1 is that you the developer need to differentiate between a NACK error some other error.    For example, Mirth may try to send the message and have a connectivity timeout.   If in fact it is due to connectivity error, the response transformer will **not** fire.

So that means, you must set the response transformer data type to RAW for you code to evaluate.  That code would need to check the response status and serialize the response to HL7 to then parse that HL7.
